### DISCUSSION TITLE
Multiple database support

### DISCUSSION THREAD

**Discussion by miguelgrinberg** [sqlalchemy/sqlalchemy#10519] [Ideas]
Hey Mike & Federico,

I'd like to run an idea I had by you, if you have a moment.

The Flask-SQLAlchemy extension has a very convenient implementation of binds that allows you to map each model to a specific bind with an added `__bind_key__` class attribute. Example:

```python
class User(db.Model):
    __bind_key__ = "auth"
    id = db.Column(db.Integer, primary_key=True)
```

These bind keys are then mapped to individual SQLAlchemy engines in the Flask configuration object. You are probably familiar with this, since it has been put in place long ago. The support of this feature is implemented in a subclass of `DeclarativeMeta` that is applied to the `db.Model` that this package uses as a declarative base.

I have implemented a similar solution in my Alchemical package, but I don't like it very much, and has proven very hard to upgrade to use the new `DeclarativeBase` introduced in 2.0. I would like to move away from the `db.Model` pattern and have a standalone `Model` class that can be recognized by type analyzers (why classes defined in attributes of an object aren't recognized for typing escapes me, but neither mypy nor pyright appear to do this).

I started experimenting with a solution to binds that does not rely on a custom metaclass, and this is what I came up with:

```python
class MetadataCollection:
    """A descriptor that maintains metadata objects for each bind and returns the
    appropriate metadata for each model."""
    metadatas = {}

    def __get__(self, obj, type):
        if type.__dict__.get('__metadata__') is None:
            bind_key = getattr(type, '__bind_key__', None)
            if bind_key not in self.metadatas:
                self.metadatas[bind_key] = MetaData()
            type.__metadata__ = self.metadatas[bind_key]
        return type.__metadata__

class Model(DeclarativeBase):
    __abstract__ = True
    metadata = MetadataCollection()
```

This solution creates a dictionary of binds, each associated with a separate `MetaData` instance. This, like a metaclass, runs while models are parsed, because the descriptor is triggered by the registry and mapper logic. The resulting `metadatas` dictionary allows me to easily sort all the tables based on their bind and generate the `binds` dictionary to pass to `Session`.

It seems to work well and feels to me like a very clean implementation. I can use this same pattern with any declarative base parent class, without having to mess with metaclasses. Do you see any issues with this approach to managing binds?

Thankful for your thoughts.

**Comment by zzzeek** [comment]
to have `__bind_key__` do something special I would use the `__init_subclass__` hook and assign `metadata` to each class as needed.  a `__get__` method is overkill.

We can take the example at https://docs.sqlalchemy.org/en/20/orm/persistence_techniques.html#simple-vertical-partitioning and add per-class metadata collections to get this same effect using idiomatic patterns:


```py
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Session

class Base(DeclarativeBase):
    pass


# use metadata / engine A
class BaseA(Base):
    __abstract__ = True
    metadata = MetaData()


# use metadata / engine B
class BaseB(Base):
    __abstract__ = True
    metadata = MetaData()


class User(BaseA):
    ...


class Address(BaseA):
    ...


class GameInfo(BaseB):
    ...


class GameStats(BaseB):
    ...


Session = sessionmaker()

Session.configure(binds={BaseA: engine1, BaseB: engine2})
```

does the above sample suit all the requirements?    you could otherwise put an `__init_subclass__` on the base or a mixin that looks for `__bind_key__` and assigns in a similar way

**Reply by miguelgrinberg** [reply]
I forgot to mention something important above, but I believe you understand this. A goal that I have is to maintain a separate `MetaData` instance per bind, basically so that I can call `create_all()`/`drop_all()` once per bind/engine, and also to allow multiple models using the same table name in different binds.

In the code above you are working with a different base per bind, which is an interesting idea that I have not considered. It's nice that it eliminates the need to add the bind name to every model. This idea would work if I could define a bind-specific base in a simpler way, like maybe:

```python
class BaseA(Base, Bind):
    __bindname__ = 'auth'
```

But this does not create a separate metadata nor makes this base abstract. I would prefer not to ask that the application developer puts SQLAlchemy specific internals such as `__abstract__` and `metadata` attributes into their application code. It would be easy to generate a base for each bind by calling a function, but I'm trying to avoid dynamically generated classes because of pyright/mypy having trouble with them. Maybe I can add an `__init_subclass__` to this `Bind` parent class I just made up. I'll play with this.

Thanks!

**Reply by CaselIT** [reply]
> Maybe I can add an `__init_subclass__` to this Bind parent class I just made up.

That should work. This would let you to create a new metadata and add the abstract to the base before calling the super `__init_subclass__ `.
You would have to distinguish between a bind base and an actual model though

**Reply by miguelgrinberg** [reply]
Okay, I was able to do this with `__init_subclass__`. In case you are curious, this is how it looks now:

```python
class UserBind(Bind, Base):
    __bind__ = 'users'

class GroupBind(Bind, Base):
    __bind__ = 'groups'

class User(UserBind):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128))

class Group(GroupBind):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128))
```

Since this is all static classes type checkers do not report any errors and provide correct completion hints for everything. I'm not sure this will be the definitive format I'll use, but I think I like it better than my own attempt. Thanks!

**Comment by zzzeek** [comment]
Sorry not sure how I hit resolve on this, that was not my intention
